File: public/code/smc__ote.ts
/**
 * OTE - Optimal Trade Entry (Fibonacci 0.62-0.79)
 */
import { Candle, Signal, makeSignal, atr, ema, swingHighs, swingLows } from "../../common";

export function detectOTESignal(
  candles: Candle[],
  opts: {
    swingLb?: number; fibLow?: number; fibHigh?: number;
    useTrendFilter?: boolean; emaPeriod?: number;
  } = {}
): Signal {
  const { swingLb = 10, fibLow = 0.62, fibHigh = 0.79,
    useTrendFilter = true, emaPeriod = 200 } = opts;
  if (candles.length < emaPeriod + 20) return makeSignal({ reason: "Insufficient data" });

  const sh = swingHighs(candles, swingLb);
  const sl = swingLows(candles, swingLb);
  const atrValues = atr(candles, 14);
  const currentAtr = atrValues[atrValues.length - 1];
  const current = candles[candles.length - 1];
  const closes = candles.map((c) => c.close);
  const emaValues = ema(closes, emaPeriod);
  const trend = current.close > emaValues[emaValues.length - 1] ? "up" : "down";

  let lastShIdx = -1, lastSlIdx = -1, lastSh = NaN, lastSl = NaN;
  for (let i = sh.length - 1; i >= 0; i--) {
    if (sh[i] !== null && lastShIdx === -1) { lastShIdx = i; lastSh = sh[i] as number; }
    if (sl[i] !== null && lastSlIdx === -1) { lastSlIdx = i; lastSl = sl[i] as number; }
    if (lastShIdx !== -1 && lastSlIdx !== -1) break;
  }
  if (isNaN(lastSh) || isNaN(lastSl)) return makeSignal({ reason: "No swing structure" });

  if (lastShIdx > lastSlIdx) {
    const impulseLow = lastSl, impulseHigh = lastSh;
    const oteLow = impulseHigh - (impulseHigh - impulseLow) * fibHigh;
    const oteHigh = impulseHigh - (impulseHigh - impulseLow) * fibLow;
    if (current.close >= oteLow && current.close <= oteHigh) {
      if (useTrendFilter && trend !== "up") return makeSignal({ reason: "OTE but trend opposite" });
      const entry = (oteLow + oteHigh) / 2;
      const stopLoss = impulseLow - 0.3 * currentAtr;
      const risk = entry - stopLoss;
      return makeSignal({
        signal: "long", entry, stop_loss: stopLoss,
        take_profit: [entry + risk * 2, entry + risk * 3, impulseHigh],
        confidence: 0.77, reason: `OTE long ${fibLow}-${fibHigh}`,
        timestamp: current.timestamp,
        metadata: { impulse_low: impulseLow, impulse_high: impulseHigh },
      });
    }
  } else {
    const impulseHigh = lastSh, impulseLow = lastSl;
    const oteLow = impulseLow + (impulseHigh - impulseLow) * fibLow;
    const oteHigh = impulseLow + (impulseHigh - impulseLow) * fibHigh;
    if (current.close >= oteLow && current.close <= oteHigh) {
      if (useTrendFilter && trend !== "down") return makeSignal({ reason: "OTE but trend opposite" });
      const entry = (oteLow + oteHigh) / 2;
      const stopLoss = impulseHigh + 0.3 * currentAtr;
      const risk = stopLoss - entry;
      return makeSignal({
        signal: "short", entry, stop_loss: stopLoss,
        take_profit: [entry - risk * 2, entry - risk * 3, impulseLow],
        confidence: 0.77, reason: `OTE short ${fibLow}-${fibHigh}`,
        timestamp: current.timestamp,
        metadata: { impulse_low: impulseLow, impulse_high: impulseHigh },
      });
    }
  }
  return makeSignal({ reason: "Price not in OTE zone" });
}

File: public/code/smc__ote.py
"""
Premium/Discount + Optimal Trade Entry (OTE)
=============================================
Son impulse hareketinin fibonacci 0.62-0.79 zone'una geri çekilme = OTE.
Premium (0.5 üstü): short fırsatı.
Discount (0.5 altı): long fırsatı.
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, atr, swing_highs_lows, ema


def detect_ote_signal(
    df: pd.DataFrame,
    swing_lb: int = 10,
    fib_low: float = 0.62,
    fib_high: float = 0.79,
    use_trend_filter: bool = True,
    ema_period: int = 200,
) -> dict:
    if len(df) < ema_period + 20:
        return make_signal(reason="Insufficient data")

    sh, sl = swing_highs_lows(df, swing_lb)
    current_atr = atr(df, 14).iloc[-1]
    trend_ema = ema(df["close"], ema_period).iloc[-1]
    current = df.iloc[-1]
    trend = "up" if current["close"] > trend_ema else "down"

    # Son swing high & low'u bul
    sh_filled = sh.ffill()
    sl_filled = sl.ffill()
    last_sh = sh_filled.iloc[-1]
    last_sl = sl_filled.iloc[-1]
    last_sh_idx = sh.last_valid_index()
    last_sl_idx = sl.last_valid_index()

    if pd.isna(last_sh) or pd.isna(last_sl):
        return make_signal(reason="No swing structure")

    # Hangisi daha yeni? Impulse yönünü belirler
    if last_sh_idx > last_sl_idx:
        # Son hareket yukarı (sl -> sh) -> long discount setup
        impulse_low = last_sl
        impulse_high = last_sh
        ote_low = impulse_high - (impulse_high - impulse_low) * fib_high
        ote_high = impulse_high - (impulse_high - impulse_low) * fib_low
        if ote_low <= current["close"] <= ote_high:
            if use_trend_filter and trend != "up":
                return make_signal(reason="OTE zone reached but trend opposite")
            entry = (ote_low + ote_high) / 2
            stop_loss = impulse_low - 0.3 * current_atr
            risk = entry - stop_loss
            return make_signal(
                signal="long",
                entry=entry,
                stop_loss=stop_loss,
                take_profit=[entry + risk * 2, entry + risk * 3, impulse_high],
                confidence=0.77,
                reason=f"OTE long: {fib_low}-{fib_high} fib zone",
                metadata={"impulse_low": impulse_low, "impulse_high": impulse_high},
            )
    else:
        # Son hareket aşağı (sh -> sl) -> short premium setup
        impulse_high = last_sh
        impulse_low = last_sl
        ote_low = impulse_low + (impulse_high - impulse_low) * fib_low
        ote_high = impulse_low + (impulse_high - impulse_low) * fib_high
        if ote_low <= current["close"] <= ote_high:
            if use_trend_filter and trend != "down":
                return make_signal(reason="OTE zone reached but trend opposite")
            entry = (ote_low + ote_high) / 2
            stop_loss = impulse_high + 0.3 * current_atr
            risk = stop_loss - entry
            return make_signal(
                signal="short",
                entry=entry,
                stop_loss=stop_loss,
                take_profit=[entry - risk * 2, entry - risk * 3, impulse_low],
                confidence=0.77,
                reason=f"OTE short: {fib_low}-{fib_high} fib zone",
                metadata={"impulse_low": impulse_low, "impulse_high": impulse_high},
            )

    return make_signal(reason="Price not in OTE zone")
